### ISSUE TITLE
Is there any plans to have support of educational version?

### ISSUE THREAD

**Issue by BondarenkoArtur** [SquidDev-CC/CC-Tweaked#784]
:wave:  Thank you for improving and supporting amazing mod!

I've seen that some teachers are using [ComputerCraftEdu](http://computercraftedu.com/) as a platform for teaching kids how to write the code. It looks pretty straight forward and now I think each kid played Minecraft at least once :sweat_smile: 
So are there any plans for supporting that mod with CC:T or at least Visual Editor that I will attach below?
![VisualEditor](http://computercraftedu.com/img/visual-editor.gif) ![CodeEditor](http://computercraftedu.com/img/code-editor.gif)

**Comment by SquidDev** [comment]
I would definitely love to see a visual editor for ComputerCraft (and generally make CC more accessible for education), but sadly it's something I'm unlikely to work on myself.

The original source of CCEdu was never released (dan200/ComputerCraft#530 and [this forum post](http://www.computercraft.info/forums2/index.php?/topic/29067-) go into a bit more detail), and so the editor would have to be reimplemented from scratch.

<details>
<summary>Some chuntering about implementation, can be ignored</summary>

Given there's less of a base to build on, I'd probably want to make some tweaks to (hopefully) make it a little more friendly. While going [full Scratch/Blockly](https://scratch.mit.edu/) probably wouldn't be right, it would be nice to make it easier to write code without getting bogged down by Lua's syntax (for instance, having to add `do`/`end`).

The problem with all of this is that, unlike TeacherGaming (the original developers), I don't really have any background in education, nor connections to teachers. This makes it much harder to test any of this, and see what changes actually help people learn.

</details>

The other big issue is that Mojang/Microsoft now have their own CCEdu-like feature in the guise of [MCEdu's agent](https://education.minecraft.net/lessons/build-with-the-agent/). It's definitely hard to know what CC would do to distinguish itself.

**TLDR:** Definitely open to making CC more accessible to new programmers and education (I know docs need a lot of work, but all suggestions are welcome!), but a visual editor is probably out of scope for now.

**Comment by TamaMcGlinn** [comment]
> the editor would have to be reimplemented from scratch.

It took a good few years to circle back to this, but I have created an open-source software stack for this, using [VSCodeWeb](https://vscode.dev/) as the code editor, so you can also use gdb as the debugger to step through code. I'm using this to teach my children and in my opinion it is much better to actually use code rather than drag and drop icons - real coders use code, and children aged 8 and up are quite capable and willing when given powerful tools.

With some *tweaks* it should be possible to run this with cc:tweaked, but since I was in the process anyway of weaning off Microsoft dependencies, the default support is for [minetest](https://minetest.org/) rather than minecraft. You can find more on [adabots.net](adabots.net) - I won't get into details here since it's kind of unrelated, but as an example, [these 14 lines implement left-hand wall following to exit a maze](https://github.com/TamaMcGlinn/AdaBots_examples/blob/fosdem2023/src/maze_solve/fernando.adb), written by [Fernando](https://github.com/Irvise) at FOSDEM'23. The [Ada library is here](https://github.com/TamaMcGlinn/AdaBots), and [minetest mod here](https://github.com/TamaMcGlinn/minetest-adabots) - PRs welcomed.

[This issue on the original ComputerCraft is also related.](https://github.com/dan200/ComputerCraft/issues/530)

**Comment by BondarenkoArtur** [comment]
> I'm using this to teach my children

Man, project is really cool. The only question is why Ada?
I saw a chunk in Ada library about why, but I'd rather teach kids Python or JS(TS). 
Never heard about Ada projects in recent time.

**Comment by TamaMcGlinn** [comment]
> why Ada?

Half of [the FAQ on AdaBots](https://adabots.net/questions) is devoted to that question. It's a good question, but you may need to push past your comfort-zone to get to the right answer. Like most programmers, you probably have a favourite language (for me, it's C++ actually), but if you're a good programmer, you should be able to agree that the best tool should be used for the job, regardless of preferences.

So, what are the criteria you would choose for a programming language to teach children with? One good criteria is industry usage, but why is that a good criteria? Because you want the children to spend their first years learning a programming language that they directly apply in real world scenarios once they are finished. That will probably be 10 years down the line. So ask yourself not just which languages are in use in industry, but which ones seem likely to last another 10 years. Ada has been in use in the industry since 1983, and companies are still discovering it and [switching to it because of its good qualities](https://www.adacore.com/papers/latitude-adopts-ada-and-spark-for-light-launcher-software-in-new-space-industry). Ada code written in 1983 can still be compiled and run today. In contrast, python code written in 2008 would need to be updated to python3 syntax, and even [the pre-release version of python today will be end-of-life in 5 years](https://devguide.python.org/versions/). JavaScript, TypeScript, rust - these are all fads compared with Ada's robust backwards compatibility.

A second important criteria is readability. Again, it might make you uncomfortable, but your personal familiarity with a certain syntax doesn't make it more readable in general. The point isn't what you, with your experience in certain languages, find quickest to understand. The point is what children actually are able to grasp and retain easiest. A key aspect of Ada syntax that helps in this is that, like Pascal, it uses keywords which are pronounced when you read the code. Second, the case-insensitivity matches your own brain's homophony - words which sound the same, *are* the same.

A third criteria for teaching is that it must be a language that the teacher understands well. If you are a dedicated teacher/parent, you have to ask yourself; "am I willing to learn a new programming language to better aid my children?" It's so much easier to just teach what you already know, and that is better than struggling to teach in a language you don't have the time to master first yourself. One day I asked my piano teacher how it's possible that he knows so many pieces so well, each time I turn the page in the book he already knows how it's played. He answered: "I don't need to know every piece in the book, I just need to stay one step ahead of you!" And that principle applies here, too. Especially if it's their first lesson, and regardless which method you end up using, it helps a lot if you privately try coding the challenge yourself before you bring your children in.

> I'd rather teach kids Python or JS(TS)

In fact, I do both Ada and python myself, because python can be found in many environments, and a programmer needs to learn multiple languages anyway. But the discussion is really "why Ada *first*" ... because it is so much easier to explain how loops work when the word "loop" is there in every loop. After that, you can refer back to it, saying "in C++ we have for-loops and while-loops", and they will understand the concept because they learned it in the easier environment first.

AdaBots can be programmed from python / js / ts ... anything that can do an HTTP call. If you want to teach using python, perhaps you would write the wrapper necessary? If you need any help, let me know.

**Comment by BondarenkoArtur** [comment]
> but if you're a good programmer, you should be able to agree that the best tool should be used for the job, regardless of preferences.

Completely agree on that. My personal language preference for prototyping is Python. But at work I have to use Java/Scala and Javascript/Typescript, and sometimes Ruby for legacy stuff. And I agree that for work scenario this stack is better even though I would enjoy more something different.

> So, what are the criteria you would choose for a programming language to teach children with?

I don't know, but I'll share my experience. My first language that I learned was Pascal, later I learned Fortran, both languages was taught by my teachers that thought it was perfect candidates. Actually I started enjoying programming when I started to learn more recent languages like Java / C#. Because: 
1. They were more practical (subjective to previous me)
2. More information was available online (books were more fresh as well)
3. Languages were more high-level, plus tools was much better (autocomplete and proper syntax highlighting)
4. Tools for creating GUI was also huge leap forward as it was allowing me to create something that I could present to non-technical people

> python code written in 2008

That's more than 10 years, and you can still install python2 if you want to just run software. 
But I would agree, migration was far from perfect. There were tools to convert python2 code to python3, and in majority cases it was working good, but was exemptions.

You mentioned that language should not be modified through the time, and even if I disagree with the point, I would still suggest to use Java if that's the case. JRE backwards compatibility is pretty good and that will allow kids after class to dive deeper into Minecraft coding for example with learning modding through Fabric. 

> the case-insensitivity matches your own brain's homophony - words which sound the same, are the same

Good point, I have to mention case sensitivity if I'm teaching and that's kinda common mistake in the beginning.

> A third criteria for teaching is that it must be a language that the teacher understands well.

Also good point, I'm not familiar with Ada, and as I'm not interested in learning it, I'm not fully understanding a selection. If you have bunch of experience with it, you will be pretty efficient. Hopefully you'll be able to focus kids on doing some personal stuff for themself and they will enjoy the process.

> programmer needs to learn multiple languages anyway

100%.

> AdaBots can be programmed from python / js / ts ... anything that can do an HTTP call. If you want to teach using python, perhaps you would write the wrapper necessary? If you need any help, let me know.

Thank you, I'll take a look. 
In any case, thanks for putting effort into teaching next generation. 🙌
